core/jvm/src/main/scala/zio/internal/WeakMap.scala

```diff
@@ -0,0 +1,242 @@
+   */
+  def isEmpty(): Boolean = {
+    garbageCollect()
+    map.isEmpty()
+  }
+
+  /**
+   * Returns the keys in the map.
+   */
+  def keySet(): Set[K] = {
+    garbageCollect()
+    val iterator = weakIterator()
+    val set      = new HashSet[K]()
+    while (iterator.hasNext) {
+      val entry = iterator.next()
+      set.add(entry.getKey)
+    }
+    set
+  }
+
+  /**
+   * Puts the specified key and value into the map, returning the old key if it
+   * exists or `null` otherwise.
+   */
+  def put(key: K, value: V): V = {
+    garbageCollect()
+    map.put(newWeakKey(key), value)
+  }
+
+  /**
+   * Puts all entries in the specified map into this map.
+   */
+  def putAll(that: Map[_ <: K, _ <: V]): Unit = {
+    garbageCollect()
+    val iterator = that.entrySet.iterator()
+    val weakMap  = new HashMap[WeakKey[K], V]()
+    while (iterator.hasNext) {
+      val entry = iterator.next()
+      weakMap.put(newWeakKey(entry.getKey), entry.getValue)
+    }
+    map.putAll(weakMap)
+  }
+
+  /**
+   * Removes the associated key from the map.
+   */
+  def remove(key: Any): V = {
+    garbageCollect()
+    map.remove(newWeakKey(key))
+  }
+
+  /**
+   * Returns the size of the map.
+   */
+  def size(): Int = {
+    garbageCollect()
+    map.size()
+  }
+
+  /**
+   * Returns all values in the map.
+   */
+  def values(): Collection[V] = {
+    garbageCollect()
+    map.values()
+  }
+
+  /**
+   * Traverses the map and removes any keys that have been garbage collected,
+   * if there is not already another simultaneous process doing the same thing.
+   */
+  private def garbageCollect(): Unit =
+    if (garbageCollecting.compareAndSet(false, true)) {
+      var key = queue.poll()
+      while (key ne null) {
+        map.remove(key)
+        key = queue.poll()
+      }
+      garbageCollecting.set(false)
+    }
+
+  /**
+   * Constructs a `java.util.Iterator` that is safe to use with a weak
+   * collection. Once `hasNext` has been called the next value, if it exists,
+   * will be captured as a strong reference, so calling `next` is guaranteed
+   * to be safe if `hasNext` returned `true`.
+   */
+  private def weakIterator(): Iterator[Entry[K, V]] =
+    new WeakIterator(map.entrySet.iterator())
+
+  /**
+   * Constructs a weak reference to the specified key with a hashcode equal to
+   * the hashcode of the key.
+   */
+  private def newWeakKey[K](key: K): WeakKey[K] =
+    new WeakKey[K](key, queue)
+}
+
+object WeakMap {
+
+  /**
+   * A `WeakKey[K]` is a weak reference to a key of type `K` with a hashcode
+   * equal to the hashcode of the key.
+   */
+  private final class WeakKey[K](key: K, queue: ReferenceQueue[_ >: K]) extends WeakReference[K](key, queue) { self =>
+    override def equals(that: Any): Boolean =
+      that match {
+        case weakKey: WeakKey[_] => self.get == weakKey.get
+        case _                   => false
+      }
+    override val hashCode: Int =
+      key.hashCode
```

Comment:
will it hold a strong reference to `key`?

Reply:
No because we eagerly compute the hash code when the `WeakKey` is created so we don't capture a reference after that. We need that hash code so we can look up the entry and remove it from the `ConcurrentHashMap` later in `garbageCollect()`. You can see if working if you run an example like [this](https://github.com/zio/zio/issues/3851#issuecomment-647207026) and add a debug `println(finalizers(mode).size)` statement in `unsafeEnsure` in `ZScope`. I will see if I can add a test for this.